contracts/MuonNodeStaking.sol

```diff
@@ -0,0 +1,624 @@
+        emit Withdrawn(msg.sender, tokenId);
+    }
+
+    /**
+     * @dev Allows users to add a Muon node.
+     * The user must have a sufficient staking amount in the BondedToken contract to run a node.
+     * @param nodeAddress The address of the Muon node.
+     * @param peerId The peer ID of the node.
+     * @param tokenId The id of the staking token.
+     */
+    function addMuonNode(
+        address nodeAddress,
+        string calldata peerId,
+        uint256 tokenId
+    ) public {
+        require(
+            users[msg.sender].tokenId == 0,
+            "You have already staked an NFT. Multiple staking is not allowed."
+        );
+
+        uint256 amount = valueOfBondedToken(tokenId);
+        require(
+            amount >= minStakeAmountPerNode,
+            "Insufficient amount to run a node."
+        );
+
+        bondedToken.transferFrom(msg.sender, address(this), tokenId);
+        users[msg.sender].tokenId = tokenId;
+
+        nodeManager.addNode(
+            nodeAddress,
+            msg.sender, // stakerAddress,
+            peerId,
+            true // active
+        );
+
+        emit MuonNodeAdded(nodeAddress, msg.sender, peerId);
+    }
+
+    /**
+     * @dev Distributes the specified reward amount to the stakers.
+     * Only callable by the REWARD_ROLE.
+     * @param reward The reward amount to be distributed.
+     */
+    function distributeRewards(uint256 reward)
+        public
+        updateReward(address(0))
+        onlyRole(REWARD_ROLE)
+    {
+        if (block.timestamp >= periodFinish) {
+            rewardRate = reward / REWARD_PERIOD;
+        } else {
+            uint256 remaining = periodFinish - block.timestamp;
+            uint256 leftover = remaining * rewardRate;
+            rewardRate = (reward + leftover) / REWARD_PERIOD;
+        }
+        lastUpdateTime = block.timestamp;
+        periodFinish = block.timestamp + REWARD_PERIOD;
+        emit RewardsDistributed(reward, block.timestamp, REWARD_PERIOD);
+    }
+
+    /**
+     * @dev Calculates the current reward per token.
+     * The reward per token is the amount of reward earned per staking token until now.
+     * @return The current reward per token.
+     */
+    function rewardPerToken() public view returns (uint256) {
+        return
+            totalStaked == 0
+                ? rewardPerTokenStored
+                : rewardPerTokenStored +
+                    (((lastTimeRewardApplicable() - lastUpdateTime) *
+                        rewardRate *
+                        1e18) / totalStaked);
+    }
+
+    /**
+     * @dev Calculates the total rewards earned by a node.
+     * @param account The staker address of a node.
+     * @return The total rewards earned by a node.
+     */
+    function earned(address account) public view returns (uint256) {
+        IMuonNodeManager.Node memory node = nodeManager.stakerAddressInfo(
+            account
+        );
+
+        if (!node.active) {
```

Comment:
This can lock tokens in the contract.
This function can calculate the total rewards. But the Muon App can sign based on the status of the node.

Reply:
`requestExit` can set `balance=0`